internal/core/model/adapters.rs

```diff
@@ -413,3 +415,510 @@ fn test_filter_model() {
+{
+    fn sort(&self, lhs: &D, rhs: &D) -> std::cmp::Ordering {
+        lhs.cmp(rhs)
+    }
+}
+
+struct FnSortHelper<F, D>
+where
+    F: FnMut(&D, &D) -> std::cmp::Ordering + 'static,
+{
+    sort_function: RefCell<F>,
+    _type: PhantomData<D>,
+}
+
+impl<F, D> SortHelper<D> for FnSortHelper<F, D>
+where
+    F: FnMut(&D, &D) -> std::cmp::Ordering + 'static,
+{
+    fn sort(&self, lhs: &D, rhs: &D) -> std::cmp::Ordering {
+        (self.sort_function.borrow_mut())(lhs, rhs)
+    }
+}
+
+struct SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    wrapped_model: M,
+    sort_helper: Box<dyn SortHelper<M::Data>>,
+    // This vector saves the indices of the elements that are not filtered out
+    mapping: RefCell<Vec<usize>>,
+    notify: ModelNotify,
+}
+
+impl<M> SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    fn build_mapping_vec(&self) {
+        self.mapping.borrow_mut().clear();
+        self.mapping.borrow_mut().extend((0..self.wrapped_model.row_count()).into_iter());
+        self.mapping.borrow_mut().sort_by(|lhs, rhs| {
+            self.sort_helper.sort(
+                &self.wrapped_model.row_data(*lhs).unwrap(),
+                &self.wrapped_model.row_data(*rhs).unwrap(),
+            )
+        });
+    }
+}
+
+impl<M> ModelChangeListener for SortModelInner<M>
+where
+    M: Model + 'static,
+{
+    fn row_changed(&self, row: usize) {
+        let removed_index = self.mapping.borrow().binary_search(&row).unwrap();
+        self.mapping.borrow_mut().remove(removed_index);
+
+        let insertion_index = if let Some(insertion_index) = self
+            .mapping
+            .borrow()
+            .iter()
+            .enumerate()
+            .skip_while(|e| {
+                self.sort_helper.sort(
+                    // inserted value
+                    &self.wrapped_model.borrow().row_data(*e.1).unwrap(),
+                    &self.wrapped_model.borrow().row_data(row).unwrap(),
+                ) == std::cmp::Ordering::Less
+            })
+            .map(|e| e.0)
+            .collect::<Vec<usize>>()
+            .first()
+        {
+            *insertion_index
+        } else {
+            self.mapping.borrow().len()
+        };
+
+        self.mapping.borrow_mut().insert(insertion_index, row);
+
+        if insertion_index == removed_index {
+            self.notify.row_changed(removed_index);
+        } else {
+            self.notify.row_removed(removed_index, 1);
+            self.notify.row_added(insertion_index, 1);
+        }
+    }
+
+    fn row_added(&self, index: usize, count: usize) {
+        if count == 0 {
+            return;
+        }
+
+        // Adjust the existing sorted row indices to match the updated source model
+        for row in self.mapping.borrow_mut().iter_mut() {
+            if *row >= index {
+                *row += count;
+            }
+        }
+
+        for row in index..(index + count) {
+            let insertion_index = if let Some(insertion_index) = self
```

Comment:
Similarly here, I think you can fetch the added row value once and then use `partition_point` to use binary search to insert, instead of linear search and the extra vector allocation.

Reply:
done.